### PR TITLE
Remove bootstrapping

### PR DESCRIPTION
**Pull request by hratoanina**
By removing the constraint of zero-initializing for the code segment of the kernel context, we can non-deterministically initialize it with the correct values, and then check them with a call to KECCAK_GENERAL. Removes ~17k CPU rows, ~70k Memory rows and the `is_bootstrap_kernel` CPU column, at the cost of one extra Memory column.

https://github.com/0xPolygonZero/plonky2/issues/1338

### CODE REVIEW

File: evm/src/generation/mod.rs [hunk 1]
```diff
@@ -149,6 +149,17 @@ fn apply_metadata_and_tries_memops<F: RichField + Extendable<D>, const D: usize>
...
         ),
+        (
+            GlobalMetadata::KernelHash,
+            KERNEL
+                .code_hash
+                .iter()
+                .enumerate()
+                .fold(0.into(), |acc, (i, &elt)| {
+                    acc + (U256::from(elt) << (224 - 32 * i))
+                }),
```
Comment (wborgeaud) [inline evm/src/generation/mod.rs:160]: Why not something like `U256::from_big_endian(&KERNEL.code_hash)`?

File: evm/src/generation/mod.rs [hunk 2]
```diff
@@ -216,6 +227,19 @@ fn apply_metadata_and_tries_memops<F: RichField + Extendable<D>, const D: usize>
     state.traces.memory_ops.extend(ops);
 }
 
+fn initialize_kernel_code<F: RichField + Extendable<D>, const D: usize>(
+    state: &mut GenerationState<F>,
+) {
+    for (i, &byte) in enumerate(KERNEL.code.iter()) {
+        let address = MemoryAddress {
+            context: 0,
+            segment: 0,
```
Comment (wborgeaud) [inline evm/src/generation/mod.rs:236]: nit: Can use `Segment::Code as usize` for clarity.

File: evm/src/memory/memory_stark.rs [hunk 3]
```diff
@@ -458,28 +463,33 @@ impl<F: RichField + Extendable<D>, const D: usize> Stark<F, D> for MemoryStark<F
...
+            // the first operation of a new address is a read, then its value must be 0.
+            let first_read_value = builder.mul_extension(next_values_limbs[i], new_first_read);
+            let first_read_user_context_constraint =
+                builder.mul_extension(next_addr_context, first_read_value);
+            yield_constr.constraint_transition(builder, first_read_user_context_constraint);
+            // For any segment outside the code segment, memory is initialized with 0. Note that the previous constraint
+            // ensure that only the code segment at context 0 is not zero-initialized.
+            let kernel_context_initializing_constraint =
+                builder.mul_extension(next_addr_segment, first_read_value);
+            yield_constr.constraint_transition(builder, kernel_context_initializing_constraint);
```
Comment (wborgeaud) [inline evm/src/memory/memory_stark.rs:491]: This is fine for now, but ideally I think we should allow non-zero initialized memory for all code segments. That way we wouldn't have to load contract code from prover inputs. Obviously we would need to take care of some security issues when doing that.

File: evm/spec/tables/memory.tex [hunk 4]
```diff
@@ -64,6 +65,15 @@ \subsubsection{Timestamps}
...
   \item Some tables other than CPU can generate memory operations, like KeccakSponge. When this happens, these operations all have the timestamp of the CPU row of the instruction which invoked the table (for KeccakSponge, KECCAK\_GENERAL).
 \end{itemize}
+
+\subsubsection{Memory initialization}
+
+By default, all memory is zero-initialized. However, to save numerous writes, we allow some specific segments to be initialized with arbitrary values.
+
+\begin{itemize}
+  \item The read-only kernel code (in segment 0, context 0) is initialized with its correct values. It's checked by hashing the segment and verifying
+that the hash value matches a verifyer-provided one.
```
Comment (wborgeaud) [inline evm/spec/tables/memory.tex:78]:
```suggestion
that the hash value matches a verifier-provided one.
```

**Review by wborgeaud — APPROVED** [review]
LGTM

**Comment by Nashtare** [comment]
You'd need to update the preprocessed ranges accordingly for `empty_txn_list` following reduction on CPU length.
In addition, it seems `log_opcode` is currently failing on this branch (not on main). It would need to be fixed before merging.

**Review by Nashtare — CHANGES_REQUESTED** [review]
You'd need to update the preprocessed ranges accordingly for `empty_txn_list` following reduction on CPU length.
In addition, it seems `log_opcode` is currently failing on this branch (not on main). It would need to be fixed before merging.

File: evm/src/memory/memory_stark.rs (hunk 3 above)
Comment (hratoanina) [reply evm/src/memory/memory_stark.rs:491]: Some must be zero-initialized for soundness (typically the main memory segment of any non-zero context), so I'd very much rather have all segments zero-initialized by default, with specific exceptions (like the code segment) made, as long as the constraints don't become too complicated and it doesn't add a prohibitive amount of extra columns.

File: evm/src/memory/memory_stark.rs [hunk 5]
```diff
@@ -325,24 +330,24 @@ impl<F: RichField + Extendable<D>, const D: usize> Stark<F, D> for MemoryStark<F
...
+                .constraint_transition(next_addr_segment * new_first_read * next_values_limbs[i]);
         }
-
-        // Check the range column: First value must be 0,
-        // and intermediate rows must increment by 1.
-        let rc1 = local_values[COUNTER];
-        let rc2 = next_values[COUNTER];
-        yield_constr.constraint_first_row(rc1);
-        let incr = rc2 - rc1;
-        yield_constr.constraint_transition(incr - P::Scalar::ONES);
```
Comment (LindaGuiga) [inline evm/src/memory/memory_stark.rs:345]: I don't think these constraints should be removed.

File: evm/src/memory/memory_stark.rs [hunk 6]
```diff
@@ -458,28 +463,33 @@ impl<F: RichField + Extendable<D>, const D: usize> Stark<F, D> for MemoryStark<F
...
         }
-
-        // Check the range column: First value must be 0,
-        // and intermediate rows must increment by 1.
-        let rc1 = local_values[COUNTER];
-        let rc2 = next_values[COUNTER];
-        yield_constr.constraint_first_row(builder, rc1);
-        let incr = builder.sub_extension(rc2, rc1);
-        let t = builder.sub_extension(incr, one);
-        yield_constr.constraint_transition(builder, t);
```
Comment (LindaGuiga) [inline evm/src/memory/memory_stark.rs:482]: I don't think these constraints should be removed.

File: evm/src/memory/memory_stark.rs (hunk 3 above)
Comment (wborgeaud) [reply evm/src/memory/memory_stark.rs:491]: Right, I was talking about the code segments, i.e., `segment=0`. If I'm correctly understanding your constraints, this allows non-zero initialized memory only for the kernel code segment `context=0, segment=0`.

Comment (hratoanina) [reply evm/src/memory/memory_stark.rs:491]: Oh yeah, sorry. Indeed, we should unconstrain the code segment for all contexts in the future.

File: evm/src/generation/mod.rs (hunk 1 above)
Comment (hratoanina) [reply evm/src/generation/mod.rs:160]: `KERNEL.code_hash` is an array of `u32`, do you have an easier way to do it?

Comment (Nashtare) [reply evm/src/generation/mod.rs:160]: IIRC, this was stored in limbs only because of KeccakSpongeStark. Now that we do the check inside the Kernel, we can change `code_hash` type to be the actual keccak output (i.e. `[u8; 32]`) and not have issues with the conversion above anymore.

File: evm/src/cpu/kernel/asm/main.asm [hunk 7]
```diff
@@ -1,5 +1,17 @@
 global main:
-    // First, initialise the shift table
+    // First, hash the kernel code
+    %mload_global_metadata(@GLOBAL_METADATA_KERNEL_LEN)
+    PUSH 0
+    DUP1
+    DUP1
```
Comment (Nashtare) [inline evm/src/cpu/kernel/asm/main.asm:6]: nit: I think it's cheaper on Memory side to call `PUSH 0` again instead of `DUP1`?

File: evm/src/memory/columns.rs [hunk 8]
```diff
@@ -35,8 +35,11 @@ pub(crate) const CONTEXT_FIRST_CHANGE: usize = VALUE_START + VALUE_LIMBS;
 pub(crate) const SEGMENT_FIRST_CHANGE: usize = CONTEXT_FIRST_CHANGE + 1;
 pub(crate) const VIRTUAL_FIRST_CHANGE: usize = SEGMENT_FIRST_CHANGE + 1;
 
+// Used to zero-initialize (or not) the memory.
+pub(crate) const NEW_FIRST_READ: usize = VIRTUAL_FIRST_CHANGE + 1;
```
Comment (Nashtare) [inline evm/src/memory/columns.rs:39]: nit: may deserve a clearer comment / var naming? A bit hard to parse without looking at the rest of the code.

**Review by Nashtare — COMMENTED** [review]
I think the failure in `log_opcode` is due to the fact that you forgot to update the memory bus with the new `GlobalMetadata` related to the Kernel on the recursive side. Not sure why it isn't caught in the `empty_txn_list` case though.

**Comment by hratoanina** [comment]
@wborgeaud I changed the zero-initializing constraint to exclude all code segments. This might make it unsound for now, but I'll work on code loading soon.

**Review by Nashtare — APPROVED** [review]
Thanks. Can you fetch latest `main` before you merge this PR? as the ranges for log_opcodes would break `main` otherwise
